language-design.md

```diff
@@ -17,7 +17,17 @@ For each quote, describe what it was about the quote that led you pick it.
+
+"I stand on this claim: I should not design a small language, and I should not design a large one. I need to design a language that can grow." [Steele, 1998, p 5]
+
+While the _Growing a Language_ article was interesting, I think the exact opposite advice applies when creating DSLs. In a general purpose programming language, growth is definitely important. You can't think of everything people will want to do with your language, and, even if you could, it would be overwhelming to include all that functionality. However, things are very different with a DSL. Since DSLs limit programmers to a specific domain, it's much easier to predict the functionality programmers will need to solve problems in that domain and to bake most of this functionality into the language from the beginning. This is highly desirable becuase people using DSLs want to work with abstractions that are natural, accurate, and relevant to them, but they don't want to (or can't) implement these abstrations themselves. While it may be useful to start off with a relatively small DSL, it will be important to have core features available from the start as programmers may not be able to add the features themselves and the DSL will likely be too limited to add the functionality from within its domain.
+
+"APIs should be easy to use and hard to misuse. It should be easy to do simple things; possible to do complex things; and impossible, or at least difficult, to do wrong things." [Bloch, 2006, p 1]
+
+This is something I'd like to know more about. It seems like almost every language gets something wrong that they later regret, but how do you avoid these kinds of mistakes in the first place? Some things are fairly straightforward; for example, Joshua Bloch suggests to "provide programmatic access to all data available in string form" and to "avoid return values that demand exceptional processing". However, some mistakes are much more subtle and hard to predict, especially when logic errors are already easy enough in general purpose programming languages. Is there an easy and/or straightforward way to design APIs, DSLs, and GPPLs so as to minimize some of the more complex errors that are common for programmers? Perhaps we could discuss this sometime in class.
```

Comment:
Super interesting questions, I think. Certainly some GPPL's like C intentionally make it easy to really mess things up in order to provide more power/flexibility to the programmer. That said I think that there isn't a "straightforward way to design" languages to avoid that, as you say in your response; I could imagine that this is the type of thing that really varies based on the context in which a language is designed.